Catch SameFileError errors when contructing RPath on Windows

johnwparent opened the pull request:
Improve error handling during link phase of Windows simulated RPaths. The symlink module, with symlinks disabled on Windows, will defer to junctions. Junctions first test if the link is viable by checking if the error already exists. For some reason in this scenario, we were previously trying to let  `shutil` copy the file, which we know will fail. This PR improves the error handling by catching the SameFileError from `shutil` and additionally, skips that code path altogether if we know we'll hit an `EEXIST` error.

In short this PR avoids an unnecessary operation, and improves the RPath error handling.

johnwparent committed:
Improve error catch from juction code path

johnwparent committed:
Improve error catch robustness

johnwparent commented:
@kwryankrattiger this should resolve your Perl issue.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -2407,7 +2407,11 @@ def _link(self, path, dest_dir):
-                if sys.platform == "win32" and e.winerror == 183:
+                if sys.platform == "win32" and (
+                    e.winerror == 183
+                    or e.errno == errno.EEXIST
```
For this check, isn't code 183 for when the file already exists? Are there other pathways on Windows that were resulting in this `errno`? AFAIK `errno` is Linux-specific so I'm surprised we'd be seeing it at all.

johnwparent replied:
We set `e.errno` explicitly in `llnl/util/symlink.py::_win32_juction` in this scenario, but in general you are correct.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -2407,7 +2407,11 @@ def _link(self, path, dest_dir):
+                if sys.platform == "win32" and (
+                    e.winerror == 183
+                    or e.errno == errno.EEXIST
+                    or isinstance(e, shutil.SameFileError)
```
It's odd that we'd be seeing this given that this is attempted inside of `if os.path.exists(dest_dir) and not dest_file == path:`. Is that on account of a lack of resolution of symlinks etc. to a normalized, absolute path?

johnwparent replied:
`not dest_file == path` will only be true if we're symlinking something like `<spack_root>/opt/path-to-package/bin/zlib.dll` to itself, as all paths are absolute. This does not help us in a scenario where we're linking something from somewhere like `<spack_root>/opt/path-to-package/x64/bin/zlib.dll` to `<spack_root>/opt/path-to-package/bin/zlib.dll`, where there will be a collision if `zlib.dll` is already in the bin as those paths are not equivalent, and that if statement will allow this through.

johnwparent replied:
We could check the destination ourselves in cases like this and avoid the error handling altogether.

scheibelp replied:
According to https://docs.python.org/3/library/shutil.html, `SameFileError` is only supposed to be raised if "source and destination in [copyfile()](https://docs.python.org/3/library/shutil.html#shutil.copyfile) are the same file."

If `<spack_root>/opt/path-to-package/bin/zlib.dll` already exists, I would not expect `SameFileError` to be raised, but rather an `IOError` with one of the already-checked error codes (e.g. `EEXIST`). Are you saying that `shutil` is raising this error in that case though?

johnwparent replied:
Yes, attached is the screenshot I've been working from to diagnose this: 
![unnamed](https://user-images.githubusercontent.com/45471568/225745753-b76cc0e3-1959-4018-a00b-1dbae59e91cb.png)

johnwparent replied:
I think from `shutil`'s perspective, they are the "same file"

scheibelp commented on lib/spack/llnl/util/symlink.py:
```diff
@@ -30,9 +30,15 @@ def symlink(real_path, link_path):
+                # we manually raised this error, we know shutil will fail the copy
+                # due to a same file error
+                # skip and report
+                raise e
```
Use `raise` rather than `raise e` (the latter interferes with the traceback)

johnwparent replied:
Huh, I did not know that... TMYK. Amended.

scheibelp commented on lib/spack/llnl/util/symlink.py:
```diff
@@ -30,9 +30,15 @@ def symlink(real_path, link_path):
+            if e.errno == errno.EEXIST:
+                # we manually raised this error, we know shutil will fail the copy
+                # due to a same file error
+                # skip and report
```
IMO it's more important that the error explains that we don't need to perform this operation, vs. explaining that we know it was raised manually (I would avoid that kind of explanation in general, including use of the term "manual")

johnwparent replied:
I'll refactor the explanation along those lines.

scheibelp reviewed (changes requested):
I have a couple questions and a couple suggestions.

johnwparent committed:
Improve symlink EEXIST error handling

johnwparent committed:
Don't check SameFileError

SameFileError cannot be encountered here as we abort prior to that
point in this scenario

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent
